### ISSUE TITLE
Clean up of spectrum/power_law.py ?

### ISSUE THREAD

**Issue by registerrier** [gammapy/gammapy#2215]
The `spectrum/power_law.py`contains a list of utility function for power law models.
In practice, only a few are actually used. And many miss a regression test. 

Are those functions useful and be properly tested or can we clean-up the file by removing a fraction of them: e.g. 
   `power_law_g_from_f`,
   `power_law_g_from_points`,
   `power_law_I_from_points`,
    `power_law_f_from_points`,
    `power_law_f_with_err`,
    `power_law_I_with_err`,
    `power_law_compatibility`
Opinions @adonath @cdeil ?

**Comment by cdeil** [comment]
Yes, this should be cleaned up.

Just some thoughts (not a proposal):
- As mentioned a the top of [gammapy/spectrum/powerlaw.py](https://github.com/gammapy/gammapy/blob/master/gammapy/spectrum/powerlaw.py) - these are Gammapy-internal helper functions, not part of the public API.
- The PL helper functions aren't much used in Gammapy at the moment, because `integrate_spectrum` and `_trapz_loglog` was added in [gammapy/spectrum/utils.py](https://github.com/gammapy/gammapy/blob/master/gammapy/spectrum/utils.py) and the PL classes in [gammapy/spectrum/models.py](https://github.com/gammapy/gammapy/blob/master/gammapy/spectrum/models.py) without using the existing PL helper functions. IMO `_trapz_loglog` is way too complex (I don't understand it), and `power_law_I_from_points` would be a better starting point to develop a good PL integrate helper function that does broadcasting and is used everywhere.
- Duplication should be removed and PL computations simplified, the question is which way to go.
- If you look at the current PL spectral classes and e.g. [PowerLaw.integral](https://docs.gammapy.org/0.12/_modules/gammapy/spectrum/models.html#PowerLaw.integral)  you see that they have most of the functionality from `powerlaw.py` already. But it's not usable from other places where PL integral is needed, e.g. [here](https://github.com/gammapy/gammapy/blob/4e9dab94152c91eac271f6a6b81d9cd4f8536707/gammapy/irf/background.py#L201) in `gammapy.irf` a PL integration is needed, but it calls `_trapz_loglog` directly because `PowerLaw` can only represent a single "bin".

So overall: I think `_trapz_loglog` and `power_law_I_from_points` / `power_law_I_with_err` should be merged and improved. Personally I prefer to have `uncertainties` handling in separate functions instead of always if-else like we have now in `_trapz_loglog` and the `PowerLaw` and other spectral model classes. Given that PL integral in energy is also used e.g. in `gammapy.irf`, we could consider moving these PL helper functions to `gammapy/utils/powerlaw.py`, but for now the cleanup could go in small steps in the existing files in `gammapy/spectrum`.

I note that in the HESS spectral software we put the same 10 years ago: generic PL helper functions grouped in a namespace (our equivalent is `gammapy.spectrum.powerlaw`), called from the spectral model methods like evaluate and integrate etc, but also from other places. Overall I think this is good, only that PL helper functions should include "piecewise PL" or "linear in log-log" in scope. The important questions to figure out is details like where to use `Quantity` or `uncertainties` objects vs. Numpy arrays. Probably by default use `Quantity`, and for `uncertainties` have separate methods that wrap the functions working with `Quantity`? Or just use Numpy arrays like we currently do in `powerlaw.py`?

@registerrier or @adonath - Can one of you make a proposal via a comment here what to do, before starting the clean-up on this? (could be a complete proposal, or just "let's start by removing / changing ABC and XYZ").

**Comment by adonath** [comment]
Thanks @cdeil, I think we could basically follow what you've written in your comment. Currently there is exactly one(!) use-case of `gammapy/spectrum/power_law.py` in the whole Gammapy code base ([here](https://github.com/gammapy/gammapy/blob/master/gammapy/spectrum/flux_point.py#L411)). So I think it's safe to either move or remove most of the functionality.

Here would be my proposal:

- Move `power_law_pivot_energy` to `PowerLaw.decorrelation_energy`
- Remove `power_law_evaluate()`, as it is available as a static method on `PowerLaw.evaluate()`
- Move `power_law_energy_flux` to a static `PowerLaw.evaluate_energy_flux()`
- Move `power_law_integral_flux` to a static `PowerLaw.evaluate_integral()`
- Remove `power_law_compatibility`, because it seems too specific.
- Remove `power_law_I_from_points` and `power_law_I_with_err` for now and make a reminder issue to simplify `_trapz_log_log`...

**Comment by adonath** [comment]
I'd be happy to prepare a PR for this...

**Comment by cdeil** [comment]
Is it a good idea to attach these as methods on `PowerLaw`?

On the one hand, yes, it would be nice to have everything in one place.
On the other hand, then we have a mix of methods that work with the instance and parameters, and others that support arrays of values e.g. for norm and don't work with the parameters.

Where would `_trapz_log_log` go?

I think my preference would be to have `_trapz_log_log` with other PL helper functions in one place, with a slight preference for that place to be `powerlaw.py` instead of attached to `PowerLaw` as static methods.

**Comment by adonath** [comment]
@cdeil The current design we have for `SpectralModel.evaluate()` is exactly the same, no? If possible the `.evaluate()` is declared as static method and then called by `__call__` in the base class with the default parameters. I think this is a good and flexible design. So why not extend it to `.evaluate_integral()` and `.evaluate_energy_flux()`? This give on the one hand the benefit of a nice API, on the other the flexibility to call the methods directly as helper methods...

**Comment by cdeil** [comment]
Fine, add it to `PowerLaw` as static methods if you like. I think partly my preference to not do that is that `models.py` is so large already, and adding the code from `powerlaw.py` and `_trapz_log_log` from `utils.py` will make it even larger.
How about doing the change to `gammapy/spectrum/models` as a package first, putting the PL class and PL helper functions in `gammapy/spectrum/models/powerlaw.py` in one file.
And then in another PR do the merging of the functionality.

A better reason to maybe keep the PL utils separate is that they might do quantity handling different, e.g. we will want to call them for background IRFs which are also approximately power-laws, and there the units are different. If you have a clear vision that unit handling will be fine and it really fits better to have these functions on the PL SpectralModel - go ahead!

I'm slightly -1 to start by deleting the PL helper functions. IMO it would be better to start by refactoring  `_trapz_log_log` to be nice and clean (IMO splitting out uncertainty handling into separate functions would be better). You might end up with something that's very similar to the current PL helper functions. Just because they weren't called by the gammapy.spectrum devs doesn't mean they are bad or useless.

**Comment by adonath** [comment]
Done in #2247.
